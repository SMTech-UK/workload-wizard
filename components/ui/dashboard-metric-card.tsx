import { ReactNode } from "react";

interface DashboardMetricCardProps {
  title: string;
  value: ReactNode;
  subtitle?: string;
  icon?: ReactNode;
  valueClassName?: string;
  children?: ReactNode; // For custom content like Progress bar
}

export function DashboardMetricCard({
  title,
  value,
  subtitle,
  icon,
  valueClassName = "",
  children,
}: DashboardMetricCardProps) {
  return (
    <div className="rounded-lg border p-6 bg-white shadow-sm">
      <div className="flex flex-row items-center justify-between space-y-0 pb-2">
        <span className="text-sm font-medium text-gray-500">{title}</span>
        {icon}
      </div>
      <div className={`text-2xl font-bold ${valueClassName}`}>{value}</div>
      {children}
      {subtitle && <div className="text-xs text-muted-foreground mt-1">{subtitle}</div>}
    </div>
  );
} 