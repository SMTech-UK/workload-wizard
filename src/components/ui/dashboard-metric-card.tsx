import { ReactNode } from "react";

interface DashboardMetricCardProps {
  title: string;
  value: ReactNode;
  subtitle?: string;
  icon?: ReactNode;
  valueClassName?: string;
  children?: ReactNode; // For custom content like Progress bar
}

export function DashboardMetricCard({
  title,
  value,
  subtitle,
  icon,
  valueClassName = "",
  children,
}: DashboardMetricCardProps) {
  return (
    <div className="rounded-lg border p-6 bg-white dark:bg-zinc-900 shadow-sm">
      <div className="flex flex-row items-center justify-between space-y-0 pb-2">
        <span className="text-sm font-medium text-gray-900 dark:text-white">{title}</span>
        {icon}
      </div>
      <div className={`text-2xl font-bold ${valueClassName}`}>{value}</div>
      {children}
      {subtitle && <div className="text-xs text-gray-600 dark:text-gray-300 mt-1">{subtitle}</div>}
    </div>
  );
} 