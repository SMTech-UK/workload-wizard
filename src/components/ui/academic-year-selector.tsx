"use client"

import React from "react";
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from "@/components/ui/select";
import { Badge } from "@/components/ui/badge";
import { useAcademicYear } from "@/hooks/useAcademicYear";
import { Calendar } from "lucide-react";

interface AcademicYearSelectorProps {
  className?: string;
  showActiveBadge?: boolean;
  onAcademicYearChange?: (academicYearId: string) => void;
}

export function AcademicYearSelector({ 
  className, 
  showActiveBadge = true,
  onAcademicYearChange 
}: AcademicYearSelectorProps) {
  const { 
    academicYears, 
    currentAcademicYear, 
    currentAcademicYearId, 
    changeAcademicYear 
  } = useAcademicYear();

  const handleAcademicYearChange = async (academicYearId: string) => {
    try {
      await changeAcademicYear(academicYearId);
      onAcademicYearChange?.(academicYearId);
    } catch (error) {
      console.error('Failed to change academic year:', error);
    }
  };

  // Function to format academic year name (e.g., "2025/26" -> "AY-25/26")
  const formatAcademicYearName = (yearName: string) => {
    // Extract the last 2 digits from each year part
    const parts = yearName.split('/');
    if (parts.length === 2) {
      const startYear = parts[0].slice(-2);
      const endYear = parts[1].slice(-2);
      return `AY-${startYear}/${endYear}`;
    }
    // Fallback to original name if format is unexpected
    return yearName;
  };

  if (!academicYears || academicYears.length === 0) {
    return (
      <div className={`flex items-center gap-2 text-muted-foreground ${className}`}>
        <Calendar className="h-4 w-4" />
        <span className="text-sm">No academic years configured</span>
      </div>
    );
  }

  return (
    <div className={`flex items-center gap-2 ${className}`}>
      <Select
        value={currentAcademicYearId || ""}
        onValueChange={handleAcademicYearChange}
      >
        <SelectTrigger className="w-auto min-w-[120px] flex items-center gap-2">
          <Calendar className="h-4 w-4 text-muted-foreground" />
          <span className="font-medium">
            {currentAcademicYear ? formatAcademicYearName(currentAcademicYear.name) : "Select AY"}
          </span>
        </SelectTrigger>
        <SelectContent>
          {academicYears.map((year: any) => (
            <SelectItem key={year._id} value={year._id}>
              <div className="flex items-center gap-2">
                <Calendar className="h-4 w-4 text-muted-foreground" />
                <span>{formatAcademicYearName(year.name)}</span>
                {year.isActive && showActiveBadge && (
                  <Badge variant="default" className="text-xs">Active</Badge>
                )}
                {year.isStaging && (
                  <Badge variant="secondary" className="text-xs">Staging</Badge>
                )}
              </div>
            </SelectItem>
          ))}
        </SelectContent>
      </Select>
    </div>
  );
} 