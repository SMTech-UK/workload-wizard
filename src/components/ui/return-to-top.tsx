"use client";

import { Button } from "@/components/ui/button";
import { ChevronUp, Home } from "lucide-react";
import { useState, useEffect } from "react";
import Link from "next/link";

export function ReturnToTop() {
  const [isVisible, setIsVisible] = useState(true); // Temporarily always visible for testing

  useEffect(() => {
    const toggleVisibility = () => {
      console.log("Scroll position:", window.pageYOffset);
      if (window.pageYOffset > 300) {
        console.log("Setting visible to true");
        setIsVisible(true);
      } else {
        console.log("Setting visible to false");
        setIsVisible(false);
      }
    };

    // Call once on mount to check initial state
    toggleVisibility();
    
    window.addEventListener("scroll", toggleVisibility);
    return () => window.removeEventListener("scroll", toggleVisibility);
  }, []);

  const scrollToTop = () => {
    console.log("Scrolling to top");
    window.scrollTo({
      top: 0,
      behavior: "smooth",
    });
  };

  console.log("ReturnToTop component render, isVisible:", isVisible);

  return (
    <>
      {/* Return to Home button - top left */}
      <Link href="/">
        <Button
          size="sm"
          variant="outline"
          className="fixed top-4 left-4 z-50 shadow-lg hover:shadow-xl transition-all duration-200"
          aria-label="Return to home"
        >
          <Home className="h-4 w-4 mr-2" />
          Home
        </Button>
      </Link>

      {/* Return to Top button - bottom left */}
      <Button
        onClick={scrollToTop}
        size="sm"
        className="fixed bottom-4 left-4 z-50 rounded-full shadow-lg hover:shadow-xl transition-all duration-200 bg-blue-500 text-white"
        aria-label="Return to top"
      >
        <ChevronUp className="h-4 w-4" />
      </Button>
    </>
  );
} 