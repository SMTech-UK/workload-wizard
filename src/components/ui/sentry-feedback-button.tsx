"use client";

import { useEffect, useState } from "react";
import * as Sentry from "@sentry/nextjs";
import { Button } from "@/components/ui/button";
import { Bug, MessageCircle } from "lucide-react";
import { useModalAwareButton } from "@/hooks/useModalAwareButton";
import { createPortal } from "react-dom";

export function SentryFeedbackButton() {
  const [feedback, setFeedback] = useState<any>(null);
  const [widget, setWidget] = useState<any>(null);
  const [mounted, setMounted] = useState(false);
  const buttonRef = useModalAwareButton();

  useEffect(() => {
    setMounted(true);
    // Get the feedback instance on the client side
    setFeedback(Sentry.getFeedback());
  }, []);

  const handleFeedbackClick = (e: React.MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();
    
    if (!feedback) return;

    if (widget) {
      // If widget is open, close it
      widget.removeFromDom();
      setWidget(null);
    } else {
      // Create and show the form immediately
      feedback.createForm().then((newWidget: any) => {
        newWidget.appendToDom();
        newWidget.open();
        setWidget(newWidget);
      }).catch((error: any) => {
        console.error('Failed to create Sentry feedback form:', error);
      });
    }
  };

  if (!feedback || !mounted) return null;

  const button = (
    <Button
      ref={buttonRef}
      variant="destructive"
      size="lg"
      onClick={handleFeedbackClick}
      onMouseDown={(e) => {
        e.preventDefault();
        e.stopPropagation();
      }}
      className="fixed bottom-4 right-4 z-[99999] shadow-lg"
      style={{
        zIndex: 99999,
        pointerEvents: 'auto',
      }}
      aria-label="Report a bug or give feedback"
    >
      <Bug className="h-4 w-4 mr-2" />
      Report an Issue
    </Button>
  );

  // Render the button in a portal to ensure it's outside modal context
  return createPortal(button, document.body);
} 