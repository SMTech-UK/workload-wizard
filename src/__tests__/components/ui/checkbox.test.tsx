import { render, screen } from '@testing-library/react';
import { Checkbox } from '@/components/ui/checkbox';

// Helper function to get elements by data-slot
const getByDataSlot = (container: HTMLElement, slot: string) => {
  return container.querySelector(`[data-slot="${slot}"]`);
};

describe('Checkbox Component - Comprehensive UI Testing', () => {
  describe('Basic Rendering and Functionality', () => {
    it('should render checkbox with default state', () => {
      const { container } = render(<Checkbox />);
      
      const checkbox = getByDataSlot(container, 'checkbox');
      expect(checkbox).toBeInTheDocument();
      expect(checkbox).toHaveAttribute('role', 'checkbox');
      expect(checkbox).toHaveAttribute('aria-checked', 'false');
    });

    it('should render checkbox with checked state', () => {
      const { container } = render(<Checkbox defaultChecked />);
      
      const checkbox = getByDataSlot(container, 'checkbox');
      expect(checkbox).toBeInTheDocument();
      expect(checkbox).toHaveAttribute('aria-checked', 'true');
    });

    it('should render checkbox indicator', () => {
      const { container } = render(<Checkbox defaultChecked />);
      
      const indicator = getByDataSlot(container, 'checkbox-indicator');
      expect(indicator).toBeInTheDocument();
      expect(indicator).toHaveClass('flex', 'items-center', 'justify-center', 'text-current');
    });
  });

  describe('Accessibility Testing', () => {
    it('should have proper ARIA attributes', () => {
      const { container } = render(<Checkbox />);
      
      const checkbox = getByDataSlot(container, 'checkbox');
      expect(checkbox).toHaveAttribute('role', 'checkbox');
      expect(checkbox).toHaveAttribute('aria-checked', 'false');
    });

    it('should have proper ARIA attributes when checked', () => {
      const { container } = render(<Checkbox defaultChecked />);
      
      const checkbox = getByDataSlot(container, 'checkbox');
      expect(checkbox).toHaveAttribute('role', 'checkbox');
      expect(checkbox).toHaveAttribute('aria-checked', 'true');
    });

    it('should support disabled state', () => {
      const { container } = render(<Checkbox disabled />);
      
      const checkbox = getByDataSlot(container, 'checkbox');
      expect(checkbox).toHaveAttribute('disabled');
      expect(checkbox).toHaveClass('disabled:cursor-not-allowed', 'disabled:opacity-50');
    });

    it('should support focus-visible styling', () => {
      const { container } = render(<Checkbox />);
      
      const checkbox = getByDataSlot(container, 'checkbox');
      expect(checkbox).toHaveClass('focus-visible:border-ring', 'focus-visible:ring-ring/50');
    });
  });

  describe('Styling and CSS Classes', () => {
    it('should have proper CSS classes for checkbox', () => {
      const { container } = render(<Checkbox />);
      
      const checkbox = getByDataSlot(container, 'checkbox');
      expect(checkbox).toHaveClass('peer', 'border-input', 'size-4', 'shrink-0', 'rounded-[4px]', 'border');
    });

    it('should have proper CSS classes for indicator', () => {
      const { container } = render(<Checkbox defaultChecked />);
      
      const indicator = getByDataSlot(container, 'checkbox-indicator');
      expect(indicator).toHaveClass('flex', 'items-center', 'justify-center', 'text-current');
    });

    it('should have proper checked state styling', () => {
      const { container } = render(<Checkbox defaultChecked />);
      
      const checkbox = getByDataSlot(container, 'checkbox');
      expect(checkbox).toHaveClass('data-[state=checked]:bg-primary', 'data-[state=checked]:text-primary-foreground');
    });
  });

  describe('Edge Cases and Error Handling', () => {
    it('should handle controlled component', () => {
      const { container } = render(<Checkbox checked />);
      
      const checkbox = getByDataSlot(container, 'checkbox');
      expect(checkbox).toHaveAttribute('aria-checked', 'true');
    });

    it('should handle indeterminate state', () => {
      const { container } = render(<Checkbox data-state="indeterminate" />);
      
      const checkbox = getByDataSlot(container, 'checkbox');
      expect(checkbox).toHaveAttribute('data-state', 'indeterminate');
    });

    it('should handle custom className', () => {
      const { container } = render(<Checkbox className="custom-class" />);
      
      const checkbox = getByDataSlot(container, 'checkbox');
      expect(checkbox).toHaveClass('custom-class');
    });
  });

  describe('Snapshot Testing for UI Consistency', () => {
    it('should match snapshot for unchecked checkbox', () => {
      const { container } = render(<Checkbox />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for checked checkbox', () => {
      const { container } = render(<Checkbox defaultChecked />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for disabled checkbox', () => {
      const { container } = render(<Checkbox disabled />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
}); 