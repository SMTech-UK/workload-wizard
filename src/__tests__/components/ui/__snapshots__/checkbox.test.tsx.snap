// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for checked checkbox 1`] = `
<button
  aria-checked="true"
  class="peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50"
  data-slot="checkbox"
  data-state="checked"
  role="checkbox"
  type="button"
  value="on"
>
  <span
    class="flex items-center justify-center text-current transition-none"
    data-slot="checkbox-indicator"
    data-state="checked"
    style="pointer-events: none;"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-check size-3.5"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M20 6 9 17l-5-5"
      />
    </svg>
  </span>
</button>
`;

exports[`Checkbox Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for default checkbox 1`] = `
<button
  aria-checked="false"
  class="peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50"
  data-slot="checkbox"
  data-state="unchecked"
  role="checkbox"
  type="button"
  value="on"
/>
`;

exports[`Checkbox Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for disabled checkbox 1`] = `
<button
  aria-checked="false"
  class="peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50"
  data-disabled=""
  data-slot="checkbox"
  data-state="unchecked"
  disabled=""
  role="checkbox"
  type="button"
  value="on"
/>
`;

exports[`Checkbox Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for invalid checkbox 1`] = `
<button
  aria-checked="false"
  aria-invalid="true"
  class="peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50"
  data-slot="checkbox"
  data-state="unchecked"
  role="checkbox"
  type="button"
  value="on"
/>
`;

exports[`Checkbox Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for multiple checkboxes 1`] = `
<div>
  <button
    aria-checked="false"
    class="peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50"
    data-slot="checkbox"
    data-state="unchecked"
    role="checkbox"
    type="button"
    value="on"
  />
  <button
    aria-checked="true"
    class="peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50"
    data-slot="checkbox"
    data-state="checked"
    role="checkbox"
    type="button"
    value="on"
  >
    <span
      class="flex items-center justify-center text-current transition-none"
      data-slot="checkbox-indicator"
      data-state="checked"
      style="pointer-events: none;"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-check size-3.5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 6 9 17l-5-5"
        />
      </svg>
    </span>
  </button>
  <button
    aria-checked="false"
    class="peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50"
    data-disabled=""
    data-slot="checkbox"
    data-state="unchecked"
    disabled=""
    role="checkbox"
    type="button"
    value="on"
  />
</div>
`;

exports[`Checkbox Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for unchecked checkbox 1`] = `
<button
  aria-checked="false"
  class="peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50"
  data-slot="checkbox"
  data-state="unchecked"
  role="checkbox"
  type="button"
  value="on"
/>
`;
